### PR TITLE
feat: Move My Expenses to Platform: Fixing Filters and Sorting 

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 45d4666</samp>

This pull request updates the codebase to use the platform v1 expense service instead of the v1 transaction service, and to add new features and filters for expenses. It also improves the code quality and readability by refactoring some files, removing unused imports, adding null checks, avoiding mutation, and renaming variables. It introduces new interfaces, enums, and methods to handle the expense filters and query parameters. It affects the files `shared-expense.service.ts`, `my-expenses.page.ts`, `my-expenses.service.ts`, `filter-state.enum.ts`, `expense-filters.model.ts`, `expense-params.model.ts`, `expenses-query-params.model.ts`, `get-expenses-query.model.ts`, and `transaction.service.ts`.

### <samp>🤖 Generated by Copilot at 45d4666</samp>

> _We are the masters of the code, we filter and we sort_
> _We use the enums and the interfaces, we make the logic short_
> _We refactor and we update, we avoid the duplication_
> _We are the platform v1, we bring the innovation_

### <samp>🤖 Generated by Copilot at 45d4666</samp>

*  Add a new enum called FilterState to define the possible values for the state filter of expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-019f13242c71e44a9677dc7df737799ddc08db71bf315166e937cf150efcc5efR1-R6))
*  Add a new interface called ExpenseFilters to extend the existing Filters interface by adding some new properties that are specific to expenses ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-537bdabc8504f79f64b66af4d7120445cff96951bd63371e94c0cad317cd9af4R1-R11))
*  Add some new properties to the existing interface ExpenseParams that are used to construct the query parameters for the expense service methods ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-3ce24505d19ea5e1c912af051d6f6610921a763d4d6b1c87cf23178199907272R10-R12))
*  Change the type of queryParams from Record<string, string> to Record<string, string | string[] | boolean> in the existing interfaces ExpensesQueryParams and GetExpenseQueryParam to allow the queryParams to accept values that are not only strings, but also arrays of strings or booleans ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-eba4df1821d93fac2802fab3ab1fa042a4db856e2c846d0c9c5527200a41111eL7-R7), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-3c4c5dcac2d1482b17d3ab6b6bdf64204c680dacb69d98946661e2dae735672cL6-R6))
*  Add some new imports and a new constructor parameter to the existing class SharedExpenseService in the file `shared-expense.service.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8a83a46f72dc182a7e19da6a6985ac4ffaee1d56ee70c0d94165d17e0a2cff98L2-R9), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8a83a46f72dc182a7e19da6a6985ac4ffaee1d56ee70c0d94165d17e0a2cff98L8-R18))
*  Add some new methods to the existing class SharedExpenseService in the file `shared-expense.service.ts` to generate the query parameters and filter pills for the expense service methods based on the filters that are passed as arguments ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8a83a46f72dc182a7e19da6a6985ac4ffaee1d56ee70c0d94165d17e0a2cff98R125-R303))
*  Remove the enum FilterState from the file `transaction.service.ts` and import it from the core enums instead ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-712edde8a2bf233766bbc21a32a8166358eca7a00851e1a25af976cff8b780b5L43-R44))
*  Add some new imports and remove some unused imports in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L1-R14), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L11-R26))
*  Change the import of FilterState from the transaction service to the core enums in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L30-R85))
*  Add a null check for the filter.type property in the generateFilterPills method in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L769-R770))
*  Change the name and type of the returned object from Partial<GetExpensesQueryParamsWithFilters> to Partial<GetExpenseQueryParam> in the addNewFiltersToParams method in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L788-R808))
*  Add a new variable called selectedFilters and assign it the value of the generateSelectedFilters method of the my-expenses.service.ts in the openFilters method in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L838-R845))
*  Add a new variable called filters1 and assign it the value of the convertFilters method of the my-expenses.service.ts in the openFilters method in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L851-R863))
*  Rename the loadData$ subject to loadExpenses$ in the clearFilters and clearText methods in the file `my-expenses.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L867-R875), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8ba341e210d765eda548d0c9e9ed2568f06e061fafd19b0fc3280e1db80c1533L875-R883))
*  Change the values of the filter.sortParam property from 'tx_amount', 'tx_txn_dt', and 'tx_org_category' to 'amount', 'spent_at', and 'category->name' respectively in the generateSortAmountPills, generateSortTxnDatePills, convertFilters, generateSelectedFilters, and generateSortCategoryPills methods in the file `my-expenses.service.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L73-R73), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L79-R79), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L89-R89), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L95-R95), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L247-R262), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L468-R473), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L485-R490), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L502-R507), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L516-R516), [link](https://github.com/fylein/fyle-mobile-app/pull/2562/files?diff=unified&w=0#diff-8f49cef762d2ab502afdcd773d9e8e4a7bef5670be85b46c0912d3da7fdf02f5L522-R522))

## Clickup
https://app.clickup.com/t/86ctv3kk4

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

**Review by jayfyle — COMMENTED** [review]
Currently card filtering does not work, have started a thread for that to figure out correct param for the field. Logic for deleting an expense is currently commented, will fix that in the next PR

**Review by OmkarJ13 — COMMENTED** [review]
Some comments

**Review by OmkarJ13 — CHANGES_REQUESTED** [review]
Added comments

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 1]
```diff
@@ -106,4 +115,183 @@ export class ExpenseService {
...
     return dialogBody;
   }
+
+  generateCardNumberParams(newQueryParams: ExpenseParams, filters: Partial<ExpenseFilters>): ExpenseParams {
+    const newQueryParamsCopy = cloneDeep(newQueryParams);
+    if (filters.cardNumbers?.length > 0) {
+      let cardNumberString = '';
+      filters.cardNumbers.forEach((cardNumber) => {
+        cardNumberString += cardNumber + ',';
+      });
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:125]: `filters.cardNumbers.join(',');` will do.

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 2]
```diff
@@ -106,4 +115,183 @@ export class ExpenseService {
...
+  generateReceiptAttachedParams(newQueryParams: ExpenseParams, filters: Partial<ExpenseFilters>): ExpenseParams {
+    const newQueryParamsCopy = cloneDeep(newQueryParams);
+    if (filters.receiptsAttached) {
+      if (filters.receiptsAttached === 'YES') {
+        newQueryParamsCopy.is_receipt_mandatory = 'is.true';
+      }
+
+      if (filters.receiptsAttached === 'NO') {
+        newQueryParamsCopy.is_receipt_mandatory = 'is.false';
+      }
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:185]: Make this `if` an `else if` for the previous `if` statement.

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 3]
```diff
@@ -106,4 +115,183 @@ export class ExpenseService {
...
+    if (filters.splitExpense) {
+      if (filters.splitExpense === 'YES') {
+        split_or_arr.push('(is_split.eq.true)');
+        newQueryParamsCopy.or = split_or_arr;
+      }
+
+      if (filters.splitExpense === 'NO') {
+        split_or_arr.push('(is_split.eq.false)');
+        newQueryParamsCopy.or = split_or_arr;
+      }
```
Comment (Chethan-Fyle) [inline src/app/core/services/platform/v1/shared/expense.service.ts:292]: Make this `if` an `else if` for the previous `if` statement.

**Review by Chethan-Fyle — APPROVED** [review]

File: src/app/core/models/platform/expense-filters.model.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { Filters } from 'src/app/fyle/my-expenses/my-expenses-filters.model';
+
+export interface ExpenseFilters extends Omit<Filters, 'state'> {
+  state: string | string[];
+  cardNumbers: string[];
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/models/platform/expense-filters.model.ts:5]: I tried changing this, but filter name comes from some other component which I was unable to locate. Furthermore, this PR is already ~300 lines and add more changes will make it difficult to review. Adding a TODO, here, I will find the exact location from where the filter name is set and change it in a separate PR, please ignore this for now 🙏

File: src/app/core/services/platform/v1/shared/shared-expense.service.ts [hunk 5]
```diff
@@ -113,4 +122,183 @@ export class SharedExpenseService {
...
+      if (filters.state.includes(FilterState.POLICY_VIOLATED)) {
+        stateOrFilter.push('and(is_policy_flagged.eq.true,or(policy_amount.is.null,policy_amount.gt.0.0001))');
+      }
+
+      if (filters.state.includes(FilterState.CANNOT_REPORT)) {
+        stateOrFilter.push('policy_amount.lt.0.0001');
+      }
+
+      if (filters.state.includes(FilterState.DRAFT)) {
+        stateOrFilter.push('state.in.(DRAFT)');
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/services/platform/v1/shared/shared-expense.service.ts]: Does not work using eq for some reason, will check further

**Review by OmkarJ13 — APPROVED** [review]

File: src/app/core/services/platform/v1/shared/expense.service.ts [hunk 6]
```diff
@@ -112,4 +121,203 @@ export class ExpenseService {
 
     return dialogBody;
   }
+
+  generateCardNumberParams(
+    newQueryParams: Record<string, string | string[] | boolean>,
+    filters: Partial<ExpenseFilters>
+  ): Record<string, string | string[] | boolean> {
+    const newQueryParamsCopy = cloneDeep(newQueryParams);
+    if (filters.cardNumbers?.length > 0) {
```
Comment (Dimple16) [inline src/app/core/services/platform/v1/shared/expense.service.ts:130]: `filters.cardNumbers?.length` should do

**Review by Dimple16 — APPROVED** [review]

**Pull request merged**